modmail/extensions/configuration_manager.py

```diff
@@ -0,0 +1,221 @@
+        kw = {}
+        kw["default"] = field.default if field.default is not marshmallow.missing else None
+        kw["name"] = field.name
+        kw["type"] = field.type
+
+        kw["metadata"] = field.metadata
+        kw["field"] = field
+
+        kw["frozen"] = field.on_setattr is attr.setters.frozen or frozen
+
+        meta: config.ConfigMetadata = field.metadata[config.METADATA_TABLE]
+        kw[config.METADATA_TABLE] = meta
+        kw["description"] = meta.description
+        kw["canonical_name"] = meta.canonical_name
+        kw["extended_description"] = meta.extended_description
+        kw["hidden"] = meta.hidden
+
+        if nested is not None:
+            kw["nested"] = nested
+
+        return cls(**kw)
+
+
+def get_all_conf_options(klass: config.ClassT, *, prefix: str = "") -> typing.Dict[str, ConfOptions]:
+    """Get a dict of ConfOptions for a designated configuration field recursively."""
+    options = dict()
+    for field in attr.fields(klass):
+        # make conf option list
+        if attr.has(field.type):
+            options.update(get_all_conf_options(field.type, prefix=prefix + field.name + "."))
+        else:
+            is_frozen = klass.__setattr__ is attr._make._frozen_setattrs
+            try:
+                conf_opt = ConfOptions.from_field(field, frozen=is_frozen, nested=prefix)
+            except KeyError as e:
+                if field.type == type:
+                    pass
+                elif config.METADATA_TABLE in e.args[0]:
+                    logger.warn(
+                        f"Issue with field '{field.name}', does not have a {config.METADATA_TABLE} key."
+                    )
+                else:
+                    logger.error(f"A key error occured with {field.name}.", exc_info=True)
+            else:
+                options[prefix + field.name] = conf_opt
+
+    return options
+
+
+class KeyConverter(commands.Converter):
+    """Convert argument into a configuration key."""
+
+    async def convert(self, ctx: Context, arg: str) -> KeyT:
+        """Ensure that a key is of the valid format, allowing a user to input other formats."""
+        # basically we're converting an argument to a key.
+        # config keys are delimited by `.`, and always lowercase, which means that we can make a few passes
+        # before actually trying to make any guesses.
+        # the problems are twofold. a: we want to suggest other keys
+        # and b: the interface needs to be easy to use.
+
+        # as a partial solution for this, `/`, `-`, `.` are all valid delimiters and are converted to `.`
+        # depending on common problems, it is *possible* to add `_` but would require fuzzy matching over
+        # all of the keys since that can also be a valid character name.
+
+        fields = get_all_conf_options(config.default().__class__)
+
+        new_arg = ""
+        for c in arg.lower():
+            if c in " /`":
```

Comment:
missing `-`

Reply:
and ` shouldn't be there..

Reply:
81c08da536125913aed657ff417dee10b91714c1